curriculum/challenges/english/03-front-end-libraries/react/use-state-to-toggle-an-element.english.md

```diff
@@ -30,6 +47,8 @@ tests:
     testString: assert.strictEqual(Enzyme.mount(React.createElement(MyComponent)).state('visibility'), false);
   - text: Clicking the button element should toggle the <code>visibility</code> property in state between <code>true</code> and <code>false</code>.
     testString: 'async () => { const waitForIt = (fn) => new Promise((resolve, reject) => setTimeout(() => resolve(fn()), 250)); const mockedComponent = Enzyme.mount(React.createElement(MyComponent)); const first = () => { mockedComponent.setState({ visibility: false }); return waitForIt(() => mockedComponent.state(''visibility'')); }; const second = () => { mockedComponent.find(''button'').simulate(''click''); return waitForIt(() => mockedComponent.state(''visibility'')); }; const third = () => { mockedComponent.find(''button'').simulate(''click''); return waitForIt(() => mockedComponent.state(''visibility'')); }; const firstValue = await first(); const secondValue = await second(); const thirdValue = await third(); assert(!firstValue && secondValue && !thirdValue); }; '
+  - text: An anonymous function should be passed to <code>setState</code>.
+    testString: const paramRegex = '[a-zA-Z$_]\\w*(,[a-zA-Z$_]\\w*)?'; const noSpaces = code.replace(/\s/g, ''); assert(new RegExp('this\\.setState\\((function\\(' + paramRegex + '\\){|([a-zA-Z$_]\\w*|\\(' + paramRegex + '\\))=>)').test(noSpaces));
```

Comment:
I'm thinking of preventing something like this:
```js
this.setState(state => ({
  visibility: !this.state.visibility
}));
```
By checking that `this` is not used until the next `}`:
```suggestion
    testString: const paramRegex = '[a-zA-Z$_]\\w*(,[a-zA-Z$_]\\w*)?'; const noSpaces = code.replace(/\s/g, ''); assert(new RegExp('this\\.setState\\((function\\(' + paramRegex + '\\){|([a-zA-Z$_]\\w*|\\(' + paramRegex + '\\))=>)(?![^}]*this)').test(noSpaces));
```

Reply:
While this does work, I'd rather we checked the presence of `this` in a separate test.  This would be trying to do too much and wouldn't help the user understand what they've done wrong.

Reply:
@ojeytonwilliams OK I've added this test in b68fe79b47c07c30d4c0b0ad0f3db71ea1be7cc3.